lib/Amazon/S3/Thin/Signer.pm

```diff
@@ -0,0 +1,100 @@
+package Amazon::S3::Thin::Signer;
+
+=head1 NAME
+
+Amazon::S3::Thin::Signer - Base class and factory for signers
+
+=head1 SYNOPSIS
+
+    # create a client object
+    my $s3client = Amazon::S3::Thin->new({
+      aws_access_key_id => $aws_access_key_id,
+      aws_secret_access_key => $secret_access_key,
+      signature_version => 4,   # optional
+    });
+
+    # create an object that can sign requests
+    my $signer = Amazon::S3::Thin::Signer->factory($s3client);
+
+    # create a request
+    my $request = HTTP::Request->new(...);
+
+    # sign the request using the client keys
+    $signer->sign_request($request);
+
+=head1 DESCRIPTION
+
+This module is used to construct an object that can sign an AWS request using
+the access keys and requested signature version from an Amazon::S3::Thin client
+object.
+
+You do not need to use this module directly; signers are created automatically
+and used to sign requests generated within the Amazon::S3::Thin module.
+
+=cut
+
+use strict;
+use warnings;
+
+=head1 CONSTRUCTORS
+
+=head2 factory($client)
+
+Constructs a signer object. C<$client> is an Amazon::S3::Thin client object.
+The signer class is based on the C<signature_version> parameter of the client
+object. Supported versions are 2 and 4.
+
+Returns the signer object.
+
+=cut
+
+sub factory
+{
+  my ($class, $thin) = @_;
+  my $version = $thin->{signature_version};
+  my $signer_class = "Amazon::S3::Thin::Signer::V$version";
+  eval "require $signer_class" or die $@;
+  $signer_class->new($thin);
+}
+
+=head2 new($client)
+
+Constructs a new object and copies all the parameters of the C<$client> object
+to the signer object. Subclasses will inherit this constructor.
+
+=cut
+
+sub new
+{
+  my ($class, $thin) = @_;
+  my $self = { %$thin };
+  bless $self, $class;
+}
+
+=head1 METHODS
+
+=head2 sign_request($request)
```

Comment:
Seeing related modules, 
http://search.cpan.org/~dbook/Net-Amazon-Signature-V4-0.18/lib/Net/Amazon/Signature/V4.pm
http://search.cpan.org/~lds/AWS-Signature4-1.02/lib/AWS/Signature4.pm
they both use a interface  `sign($request)` ,  so I prefer it.